### ISSUE TITLE
Error chaincode is being launched

### ISSUE THREAD

**Issue by ki0apa** [hyperledger/composer#1670]
I've sent around 500 REST requests to a network and several errors occur.

## Context

I set up a basic network using the tutorial found here: [Developer Guide](https://hyperledger.github.io/composer/tutorials/developer-guide.html). I then made a REST server and spammed it with requests. These requests either created a new asset or participant or submitted a transaction. After submitting about 500 requests, the network seems to crash and several errors are given. If I followed the steps on the website correctly, this should not be occurring. I am not sure if it is an error with the website instructions or composer itself.

## Expected Behavior
I should be able to post many requests without any errors.

## Actual Behavior

I'm getting some of the following errors. 

`Unhandled error for request POST /api/org.example.mynetwork.Trader: Error: error trying to query chaincode. Error: Error executing chaincode: Failed to execute transaction (Timeout expired while executing transaction) at channel.queryByChaincode.then.catch (/home/ubuntu/composer/my-network/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:758:34)`

`Unhandled error for request POST /api/org.example.mynetwork.Commodity: Error: error trying to query chaincode. Error: Error executing chaincode: premature execution - chaincode (my-network:0.9.2) is being launched at channel.queryByChaincode.then.catch (/home/ubuntu/composer/my-network/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:758:34)`

In the peer docker logs, the first error reported is:

` [chaincode] processStream -> ERRO 69f63[0m Error handling chaincode support stream: rpc error: code = Canceled desc = context canceled`

## Steps to Reproduce
1. Download the basic sample network using the link provided in the context section on an aws instance (Ubuntu 16.04, t2.xlarge, Ohio)
2. Set up a rest server on the aws instance.
3. Send up to 500 requests to the server.

## Existing issues

- [x] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [ ] [GitHub Issues](https://github.com/hyperledger/composer/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

- [Timeout Error](https://github.com/hyperledger/composer/issues/1514) 
- I asked this question on rocket chat and stack overflow and got no sufficient response.

## Context
I'm trying to test the scalability of Hyperledger Composer.

## Your Environment

Docker version 17.06.0-ce, build 02c1d87
docker-compose version 1.13.0, build 1719ceb
docker-py version: 2.2.1
CPython version: 2.7.13
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
composer-cli                   v0.10.0
composer-admin                 v0.10.0
composer-client                v0.10.0
composer-common                v0.10.0
composer-runtime-hlf           v0.10.0
composer-runtime-hlfv1         v0.10.0
npm-version: 5.3.0
node version: 6.11.1

The rest server is running on an aws instance and I am sending requests on a local computer with fedora 25.

**Comment by davidkel** [comment]
@ki0apa Thanks for the Peer logs. It looks like the chaincode container for some reason terminates. Fabric on the next request notices this and brings up another chaincode container and isn't ready to accept requests. Unfortunately the chaincode logs you attached are for the newly spawned container and so don't provide any insight into why the chaincode container terminated. Could you add the following environment variable to your Peer definition and restart your Peer (or fabric) and run the test again. This will put the chaincode container logs into the Peer logs so will hopefully capture information from the failing chaincode container.
`CORE_VM_DOCKER_ATTACHSTDOUT=true`

So for example if you use docker-compose to bring up the environment then your peer definition might start with something like
```
  peer0.org1.example.com:
    container_name: peer0.org1.example.com
    image: hyperledger/fabric-peer:x86_64-1.0.0
    environment:
      - CORE_VM_DOCKER_ATTACHSTDOUT=true
      - CORE_LOGGING_PEER=debug
      - CORE_CHAINCODE_LOGGING_LEVEL=DEBUG
      - CORE_VM_ENDPOINT=unix:///host/var/run/docker.sock
      - CORE_PEER_ID=peer0.org1.example.com
      - CORE_PEER_ADDRESS=peer0.org1.example.com:7051
      - CORE_VM_DOCKER_HOSTCONFIG_NETWORKMODE=composer_default
      - CORE_PEER_LOCALMSPID=Org1MSP
      - CORE_PEER_MSPCONFIGPATH=/etc/hyperledger/peer/msp
      - CORE_LEDGER_STATE_STATEDATABASE=CouchDB
      - CORE_LEDGER_STATE_COUCHDBCONFIG_COUCHDBADDRESS=couchdb:5984
    working_dir: /opt/gopath/src/github.com/hyperledger/fabric
    command: peer node start --peer-defaultchain=false
```

**Comment by ki0apa** [comment]
Th

**Comment by ki0apa** [comment]
Sorry about closing this thread, that was an accident.

I was originally launching my peer using this command:

`docker exec -e "CORE_PEER_MSPCONFIGPATH=/etc/hyperledger/msp/users/Admin@org1.example.com/msp" peer0.org1.example.com peer channel join -b composerchannel.b\
lock`

so I added this segment of code after docker exec:

` -e "CORE_VM_DOCKER_ATTACHSTDOUT=true"`

I added my updated peer logs to the [github](https://github.com/ki0apa/Composer_Error). (I had to split the peer_err.txt into two files because it exceed GitHub's file limit).

**Comment by ki0apa** [comment]
Just realized my solution above would not work. I found the docker-compose.yml file and changed it from there. Updates are in the github.

**Comment by dselman** [comment]
Lots of CouchDB errors in those logs, including this one.
```
2017-07-26 17:36:08.230 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO 8dc38[0m [36m2017-07-26 17:36:08.230 UTC [Composer] Debug -> DEBU 2ffa3[0m @JS : BusinessNetworkDefinition:fromArchive()             > 5641
[36m2017-07-26 17:36:08.230 UTC [couchdb] handleRequest -> DEBU 8dc39[0m Couch DB Error:not_found,  Status Code:404,  Reason:missing
[36m2017-07-26 17:36:08.230 UTC [couchdb] ReadDoc -> DEBU 8dc3a[0m Document not found (404), returning nil value instead of 404 error
[36m2017-07-26 17:36:08.230 UTC [chaincode] func1 -> DEBU 8dc3b[0m [640fe3a9]No state associated with key: Asset:org.hyperledger.composer.system.Identity7d195e0cbc77f47d78f39e21ec879210cdff2f8506d5f326af245563ebfcbb69. Sending RESPONSE with an empty payload
```

**Comment by dselman** [comment]
Are you using Composer queries? If not, it might be worth switching to use LevelDB and see if you can reproduce this.

**Comment by davidkel** [comment]
So the chaincode container definitely dies, unfortunately the log info doesn't provide any insight into why. Here is a relevant section of the log
```
2017-07-26 17:36:58.970 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e4228[0m [36m2017-07-26 17:36:58.970 UTC [Composer] Debug -> DEBU 4d1aa[0m Exiting NewDataCollection &{0xc422201348 0xc421c73800 $sysdata}
2017-07-26 17:36:58.971 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e4229[0m [36m2017-07-26 17:36:58.971 UTC [Composer] Debug -> DEBU 4d1ad[0m Exiting DataService.getCollection 0
2017-07-26 17:36:58.971 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e422a[0m [36m2017-07-26 17:36:58.971 UTC [Composer] Debug -> DEBU 4d1ae[0m @JS : Context                  :loadBusinessNetworkRecord() Getting business network archive from the $sysdata collection 
[36m2017-07-26 17:36:58.972 UTC [chaincode] processStream -> DEBU e422c[0m [cc86c956]Received message GET_STATE from shim
[36m2017-07-26 17:36:58.972 UTC [chaincode] HandleMessage -> DEBU e422d[0m [cc86c956]Fabric side Handling ChaincodeMessage of type: GET_STATE in state ready
[36m2017-07-26 17:36:58.972 UTC [chaincode] afterGetState -> DEBU e422e[0m [cc86c956]Received GET_STATE, invoking get state from ledger
[36m2017-07-26 17:36:58.972 UTC [chaincode] filterError -> DEBU e422f[0m Ignoring NoTransitionError: no transition
[36m2017-07-26 17:36:58.972 UTC [couchdb] handleRequest -> DEBU e422b[0m Exiting handleRequest()
[36m2017-07-26 17:36:58.972 UTC [chaincode] func1 -> DEBU e4230[0m [cc86c956] getting state for chaincode my-network, key $sysdatabusinessnetwork, channel composerchannel
[36m2017-07-26 17:36:58.972 UTC [statecouchdb] GetState -> DEBU e4232[0m GetState(). ns=my-network, key=$sysdatabusinessnetwork
2017-07-26 17:36:58.972 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e4234[0m [36m2017-07-26 17:36:58.971 UTC [Composer] Debug -> DEBU 4d1af[0m Entering DataCollection.get &{0xc4213747a0}
2017-07-26 17:36:58.972 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e4235[0m [36m2017-07-26 17:36:58.971 UTC [Composer] Debug -> DEBU 4d1ab[0m Entering NewDataCollection &{0xc4214743c0} &{0xc421b4b3c8 0xc421802980} 0xc421b3b5f0
[36m2017-07-26 17:36:58.972 UTC [couchdb] ReadDoc -> DEBU e4231[0m Exiting ReadDoc()
[36m2017-07-26 17:36:58.972 UTC [chaincode] func1 -> DEBU e4236[0m [7a56680c]Got state. Sending RESPONSE
[36m2017-07-26 17:36:58.972 UTC [chaincode] 1 -> DEBU e4237[0m [7a56680c]handleGetState serial send RESPONSE
[36m2017-07-26 17:36:58.972 UTC [couchdb] ReadDoc -> DEBU e4233[0m Entering ReadDoc()  id=[my-network$sysdatabusinessnetwork]
[36m2017-07-26 17:36:58.972 UTC [couchdb] handleRequest -> DEBU e4238[0m Entering handleRequest()  method=GET  url=http://couchdb:5984/composerchannel/my-network%00%00$sysdata%00businessnetwork%00?attachments=true
[36m2017-07-26 17:36:58.973 UTC [couchdb] handleRequest -> DEBU e4239[0m HTTP Request: GET /composerchannel/my-network%00%00$sysdata%00businessnetwork%00?attachments=true HTTP/1.1 | Host: couchdb:5984 | User-Agent: Go-http-client/1.1 | Accept: multipart/related | Accept-Encoding: gzip |  | 
2017-07-26 17:36:58.973 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e423a[0m [36m2017-07-26 17:36:58.973 UTC [Composer] Debug -> DEBU 4d1b0[0m Exiting NewDataCollection &{0xc421b9f7d8 0xc421802980 $sysdata}
2017-07-26 17:36:58.973 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e423b[0m [36m2017-07-26 17:36:58.971 UTC [Composer] Debug -> DEBU 4d1ac[0m Entering NewDataCollection &{0xc4203f9440} &{0xc4202800c0 0xc421c73b00} 0xc421b3b730
2017-07-26 17:36:58.974 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e423c[0m [36m2017-07-26 17:36:58.974 UTC [Composer] Debug -> DEBU 4d1b1[0m Exiting NewDataCollection &{0xc421b9f748 0xc421c73b00 $sysdata}
2017-07-26 17:36:58.974 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e423d[0m [36m2017-07-26 17:36:58.974 UTC [Composer] Debug -> DEBU 4d1b2[0m Exiting DataService.getCollection 0
2017-07-26 17:36:58.974 UTC [dev-peer0.org1.example.com-my-network-0.10.0] func2 -> INFO e423e[0m [36m2017-07-26 17:36:58.974 UTC [Composer] Debug -> DEBU 4d1b3[0m Entering NewDataCollection &{0xc42183a440} &{0xc421707f50 0xc4212da680} 0xc421c30230
[36m2017-07-26 17:36:58.975 UTC [couchdb] handleRequest -> DEBU e423f[0m Exiting handleRequest()
[36m2017-07-26 17:36:58.976 UTC [couchdb] ReadDoc -> DEBU e4240[0m Exiting ReadDoc()
[36m2017-07-26 17:36:58.976 UTC [chaincode] func1 -> DEBU e4241[0m [cc86c956]Got state. Sending RESPONSE
[36m2017-07-26 17:36:58.976 UTC [chaincode] 1 -> DEBU e4242[0m [cc86c956]handleGetState serial send RESPONSE
[36m2017-07-26 17:36:58.976 UTC [couchdb] handleRequest -> DEBU e4243[0m Exiting handleRequest()
[36m2017-07-26 17:36:58.977 UTC [couchdb] ReadDoc -> DEBU e4244[0m Exiting ReadDoc()
[36m2017-07-26 17:36:58.977 UTC [chaincode] func1 -> DEBU e4245[0m [9c4fac3c]Got state. Sending RESPONSE
[36m2017-07-26 17:36:58.977 UTC [chaincode] 1 -> DEBU e4246[0m [9c4fac3c]handleGetState serial send RESPONSE
[36m2017-07-26 17:36:58.977 UTC [couchdb] handleRequest -> DEBU e4247[0m Exiting handleRequest()
[36m2017-07-26 17:36:58.977 UTC [couchdb] ReadDoc -> DEBU e4248[0m Exiting ReadDoc()
[36m2017-07-26 17:36:58.977 UTC [chaincode] func1 -> DEBU e4249[0m [5744c318]Got state. Sending RESPONSE
[36m2017-07-26 17:36:58.977 UTC [chaincode] 1 -> DEBU e424a[0m [5744c318]handleGetState serial send RESPONSE
[31m2017-07-26 17:36:59.061 UTC [chaincode] processStream -> ERRO e424b[0m Error handling chaincode support stream: rpc error: code = Canceled desc = context canceled
[36m2017-07-26 17:36:59.062 UTC [chaincode] deregisterHandler -> DEBU e424c[0m Deregister handler: my-network:0.10.0
[36m2017-07-26 17:36:59.062 UTC [chaincode] deregisterHandler -> DEBU e424d[0m Deregistered handler with key: my-network:0.10.0
2017-07-26 17:36:59.076 UTC [dockercontroller] func2 -> INFO e424e[0m Container dev-peer0.org1.example.com-my-network-0.10.0 has closed its IO channel

```

**Comment by davidkel** [comment]
@ki0apa could you attach your client code so that we can try to recreate it please ?

**Comment by ki0apa** [comment]
Here is the Client Code:

`#!/usr/bin/env python

import random
import requests
import time
import threading

class rest:

	def __init__(self, file):
		self.commodities = 0
		self.traders = 0
		self.IP = "52.15.130.30"
		self.name = self.randomString()
		self.postRandomTrader()
		self.postRandomCommidity()
		while True:
			self.name = self.randomString()
			time = self.timeInMilli()
			rand = random.randint(1, 2)
			if rand == 1:
				self.postRandomCommidity()
			elif rand == 2:
				self.postRandomTrader()
			else:
				self.postRandomTrade()
			file.write(str(self.timeInMilli() - time) + "\n")
			file.flush()

	def ping(self):
		print requests.get("http://" + self.IP + ":3000/api/system/ping", headers={"Accept": "application/json"}).text

	def timeInMilli(self):
		return int(round(time.time() * 1000))

	def randomString(self):
		return "".join(str(chr(random.randint(0,25) + ord('a'))) for x in range(10))

	def decode(self, json):
		return "{" + (",".join("\"%s\": \"%s\"" % (k,v) for k,v in json.items())) + "}"

	def postRandomCommidity(self):
		self.ping()
		try:
			r = requests.post("http://" + self.IP + ":3000/api/org.example.mynetwork.Commodity", headers={"Accept": "application/json", "Content-Type": "application/json"}, data=self.decode({
			   "$class": "org.example.mynetwork.Commodity", 
			   "tradingSymbol": self.name + "_c" + str(self.commodities), 
			   "description": "string", 
			   "mainExchange": "string",  
			   "quantity": 0, 
		   		"owner": "resource:org.example.mynetwork.Trader#" + self.name + "_t" + str(random.randint(0, self.traders - 1))
	   		}), timeout=5)
	   	except Exception:
			print "fail"
	   		pass
	   	else:
	   		print r.text
	   		self.commodities += 1


	def postRandomTrader(self):
		self.ping()
		try:
			r = requests.post("http://" + self.IP + ":3000/api/org.example.mynetwork.Trader", headers={"Accept": "application/json", "Content-Type": "application/json"}, data=self.decode({
				"$class": "org.example.mynetwork.Trader",
				"tradeId": self.name + "_t" + str(self.traders),
				"firstName": "string",
				"lastName": "string"
		    }), timeout=5)
		except Exception:
			print "fail"
			pass
		else:
	   		print r.text
			self.traders += 1

	def postRandomTrade(self):
		self.ping()
		try:
			r = requests.post("http://" + self.IP + ":3000/api/org.example.mynetwork.Trade", headers={"Accept": "application/json", "Content-Type": "application/json"}, data=self.decode({
				"$class": "org.example.mynetwork.Trade",
		  		"commodity": "resource:org.example.mynetwork.Commodity#" + self.name + "_c" + str(random.randint(0, self.commodities - 1)),
		  		"newOwner": "resource:org.example.mynetwork.Trader#" + self.name + "_t" + str(random.randint(0, self.traders - 1)) 
		    }), timeout=5)
		except Exception:
			print "fail"
			pass
		else:
			print r.text

threads = []
file = open("log.csv", "w")
for i in range(10):
    t = threading.Thread(target=rest, args=(file,))
    threads.append(t)
    t.start()

`

You can replace the IP with the public IP of the AWS instance you installed composer on.

**Comment by davidkel** [comment]
It's a memory leak, continuing investigation

**Comment by ki0apa** [comment]
Yes, I figured this out already. Sorry, for some reason I thought I posted this. The docker container of the network leaks after it exceeds its 2G memory limit, this is because memory swapping is not enabled on AWS instances.

**Comment by davidkel** [comment]
@ki0apa Thanks for the script. Now comes the fun part of trying to track down the leak :-)

**Comment by davidkel** [comment]
Found the cause of a bigger memory leak, but there could be a memory problem related to the number of assets that have been created. Which requires more investigation.

**Comment by ki0apa** [comment]
Great! Is it possible you can commit these changes to a branch or show me how to fix the memory leak so I can continue working on my project? Thanks for the help.

**Comment by davidkel** [comment]
What I believe is happeningon the bigger memory leak issue is this. The composer runtime keeps a pool of 8 Composer objects as they are expensive to create. However if more than 8 concurrent requests come in then the composer runtime dynamically creates extra as required. However if they cannot be returned to the pool as the pool has a fixed size limit, they are just thrown away. I don't believe the memory allocated by and for DukTape is being freed and this is the memory leak but I am still trying to definitely prove this. As a test I increased the Pool to 15 and then the memory stays stable for a lot longer (but unfortunately still increases) and when it finally runs out of memory and crashes on restart the container starts up with a near maximum memory utilisation so need to understand the cause of this other issue.

**Comment by davidkel** [comment]
So final update for the day, after experimenting with different solutions. If I restrict the pool to 8 and fix it at that, the memory still creeps up but much more slowly. When I ran a test where no pool was used and composer objects were created and trashed (including the duktape heap) then memory stabilised which could suggest there is another small memory leak within duktape itself, but not conclusive. Interestingly trashing the duktape context should that although a composer object was finished with from the Go layer point of view in fact there was at least one timer that would fire later and it would be useful to know what that timer was to get a complete picture as if trashing the context is the only way to contain the leak then maybe we could have a policy to trash a Composer object after a certain number of uses but only if there is no timer to pop

**Comment by davidkel** [comment]
So I have a fix for the Big memory leak, awaiting review approval.

**Comment by ki0apa** [comment]
Thanks for the help!

**Comment by davidkel** [comment]
@ki0apa thanks for the testing and the stress test script. The fix should now be in our published unstable builds for release in 0.10.2. There is a follow-on investigation as to why memory still appears to creep up slowly see #1721, but they could be expected as the asset count increases.
